### PR TITLE
Added id attribute to IpcMessage

### PR DESCRIPTION
**Pull request by MattTaylorDLS**
Includes login in IpcClient python class to discard messages where it is
expecting a reply and the reply id does not match the request id

#86
BC-625

### CODE REVIEW

File: tools/python/odin_data/ipc_client.py [hunk 1]
```diff
@@ -22,26 +22,37 @@ def __init__(self, ip_address, port):
...
         self.ctrl_channel = IpcChannel(IpcChannel.CHANNEL_TYPE_DEALER)
         self.ctrl_channel.connect(self.ctrl_endpoint)
+        
+        self.message_id = 0
 
         self._lock = RLock()
 
     def _send_message(self, msg, timeout):
+        msg.set_msg_id(self.message_id)
+        self.message_id = (self.message_id + 1) % 4294967296
```
Comment (timcnicholls) [inline tools/python/odin_data/ipc_client.py]: Probably shouldn't hard code a magic number here. Maybe define a class level constant that evaluates 2**32 so it is more obvious why this is happening?

**Comment by timcnicholls** [comment]
@ajgdls I missed it from the member initialisation so it's technically undefined (although probably defaults to false) at startup. It's only use is in reporting back to the client once the controller is running its event loop, by which time configure() has been called, even with only a partial configuration, so it has been set before any client can read it. The gcc version on RHEL6 doesn't complain, but later ones might.

**Comment by MattTaylorDLS** [comment]
Yes, I changed it to get rid of the compiler warning (only one in the code i think). I was making a change in that file anyway, so thought i'd get rid of that warning at the same time

File: common/src/IpcMessage.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@ IpcMessage::IpcMessage(MsgType msg_type, MsgVal msg_val, bool strict_validation)
     strict_validation_(strict_validation),
     msg_type_(msg_type),
     msg_val_(msg_val),
+    msg_id_(0),
```
[the remark this answers is not in the document]
Comment (MattTaylorDLS) [reply common/src/IpcMessage.cpp]: No 0 is as valid as any other number. The msg_id is an unsigned int so we wouldn't be able to set it to -1.

This does mean that you could get the case where a default initialised message appears to have the same id as the request that was sent, if the request had the id of 0. This was one of the things I was going to discuss on Thursday

**Pull request merged**